backend/metadata_writer/src/metadata_writer.py

```diff
@@ -83,16 +83,7 @@ def output_name_to_argo(name: str) -> str:
 
 
 def argo_artifact_to_uri(artifact: dict) -> str:
-    if 's3' in artifact:
-        s3_artifact = artifact['s3']
-        return 'minio://{bucket}/{key}'.format(
-            bucket=s3_artifact.get('bucket', ''),
-            key=s3_artifact.get('key', ''),
-        )
-    elif 'raw' in artifact:
-        return None
-    else:
-        return None
+    return 'argo-artifact://' + json.dumps(artifact, sort_keys=True)
```

Comment:
Discussion: is this format following any existing convention? My gut feeling is we can just use `json.dumps(artifact, sort_keys=True)` without the prefix.

Did I miss any context?

Reply:
There is no existing convention. But this is supposed to be a URI, so I think it's better to add a URI schema that specifies the "URI" format.
Imagine there was a different non-Argo backend. In that case we'd use a different schema and the format would be different.

Reply:
I see, that makes sense to me.

Reply:
Following is an artifact example,  in metadata writer, we can definitely dump the whole object and send back to caller. However, we still need to parse object (in this case, get artifact.endpoint to infer the right storage) at the frontend end if there's any customization on the caller side.  (for example, frontend renders different links for artifact file based on object storage provider. see #3531). 

```
"artifacts": [
    {
      "name": "mlpipeline-ui-metadata",
      "path": "/mlpipeline-ui-metadata.json",
      "s3": {
        "endpoint": "s3.amazonaws.com",
        "bucket": "jiaxin-kubeflow-pipeline-data",
        "insecure": false,
        "accessKeySecret": {
          "name": "aws-secret",
          "key": "AWS_ACCESS_KEY_ID"
        },
        "secretKeySecret": {
          "name": "aws-secret",
          "key": "AWS_SECRET_ACCESS_KEY"
        },
        "key": "artifacts/iris-classification-pipeline-9shz4/iris-classification-pipeline-9shz4-385916632/mlpipeline-ui-metadata.tgz"
      },
      "optional": true
    }
  ]

```

Reply:
@Jeffwan 
>However, we still need to parse object (in this case, get artifact.endpoint to infer the right storage) at the frontend end if there's any customization on the caller side.

That's true.
The idea is to remove the Metadata Writer "bottleneck" between Argo and Frontend. For example, today, if someone wanted to make clickable HDFS artifacts, they'll have to modify both Metadata Writer and Frontend. After this change, they'll never have to modify Metadata Writer as they'll always have the whole information.
Also I remember that even with S3 there were some inconveniences due to, say accessKeySecret and secretKeySecret not being available.

Reply:
I thought a bit more and decided to move the func artifact information to a custom property. Forgot about those. It will still be available to the Frontend without any weird URIs.